### PR TITLE
DEPR: Series.ptp()

### PR DESCRIPTION
**Pull request by KalyanGokhale**
xref #18262
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: doc/source/whatsnew/v0.24.0.txt [hunk 1]
```diff
@@ -83,7 +83,7 @@ Deprecations
 ~~~~~~~~~~~~
 
 - :meth:`DataFrame.to_stata`, :meth:`read_stata`, :class:`StataReader` and :class:`StataWriter` have deprecated the ``encoding`` argument.  The encoding of a Stata dta file is determined by the file type and cannot be changed (:issue:`21244`).
--
+- :meth:`Series.ptp` is deprecated. Use numpy.ptp instead (:issue:`21614`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.txt]: you can put a reference to ``numpy.ptp`` as well

File: pandas/core/generic.py [hunk 2]
```diff
@@ -8871,8 +8871,15 @@ def _add_series_only_operations(cls):
         axis_descr, name, name2 = _doc_parms(cls)
 
         def nanptp(values, axis=0, skipna=True):
+            """
+            .. deprecated:: 0.24.0
+                Use numpy.ptp instead
```
Comment (jreback) [inline pandas/core/generic.py]: I am not sure this is the right format for the doc-str, cc @jorisvandenbossche

File: pandas/tests/series/test_analytics.py [hunk 3]
```diff
@@ -1381,8 +1381,12 @@ def test_ptp(self):
         ser = Series(arr)
         assert np.ptp(ser) == np.ptp(arr)
 
-        # GH11163
         s = Series([3, 5, np.nan, -3, 10])
+
+        # GH21614
```
Comment (jreback) [inline pandas/tests/series/test_analytics.py]: make this another method. you need to catch any warnings for calling ``s.ptp()`` anywhere in the codebase, or just call ``np.ptp``

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/generic.py (hunk 2 above)
Comment (KalyanGokhale) [reply pandas/core/generic.py]: Thanks - will wait for confirmation on right format from @jorisvandenbossche 

Based on the other deprecation PRs I checked, this seemed to be the expected format - for other messages I had noted earlier I got Travis failures (example below from different PR submitted)
```python
Check for deprecated messages without sphinx directive
pandas/core/indexes/multi.py:        DEPRECATED. to_hierarchical will be removed in a future version.
Check for deprecated messages without sphinx directive DONE
...
The command "ci/lint.sh" exited with 1.
```

Comment (jorisvandenbossche) [reply pandas/core/generic.py]: Yes, this is the correct format. But, shouldn't this rather be in `cls.ptp` docstring passed to `_make_stat_function` below? 
(didn't really check, but make sure that `pd.Series.ptp?` gives the correct docstring)

Comment (jorisvandenbossche) [reply pandas/core/generic.py]: Did a quick check, and indeed adding the deprecation message does not update the actual docstring of `Series.ptp`, you need to add it to the argument passed to `_make_stat_function` a few lines below

**Review by jorisvandenbossche — CHANGES_REQUESTED** [review]

**Comment by KalyanGokhale** [comment]
Thanks @jorisvandenbossche have moved the docstring to `_make_stat_function`

File: doc/source/whatsnew/v0.24.0.txt (hunk 1 above)
Comment (KalyanGokhale) [reply doc/source/whatsnew/v0.24.0.txt]: done - updated for all comments Thanks

**Review by jreback — APPROVED** [review]

**Review by jreback — CHANGES_REQUESTED** [review]
can you remove from doc/api.rst as well, otherwise lgtm.

**Comment by KalyanGokhale** [comment]
@jreback thanks - done - all green

**Review by jreback — APPROVED** [review]

File: pandas/core/generic.py [hunk 4]
```diff
@@ -8875,13 +8875,20 @@ def _add_series_only_operations(cls):
...
+                          FutureWarning, stacklevel=4)
             return nmax - nmin
 
         cls.ptp = _make_stat_function(
             cls, 'ptp', name, name2, axis_descr,
-            """Returns the difference between the maximum value and the
+            """
+            .. deprecated:: 0.24.0
+                Use numpy.ptp instead
+            Returns the difference between the maximum value and the
```
Comment (jorisvandenbossche) [inline pandas/core/generic.py]: This needs a blank line above this one. Can you also put the deprecated part below the text?

Comment (KalyanGokhale) [reply pandas/core/generic.py]: Thanks - Done - let me know if the deprecated note below main text needs any edits, have also added the line `Use numpy.ptp instead` again

File: pandas/tests/series/test_analytics.py [hunk 5]
```diff
@@ -1395,34 +1395,44 @@ def test_numpy_argmax_deprecated(self):
...
         N = 1000
         arr = np.random.randn(N)
         ser = Series(arr)
         assert np.ptp(ser) == np.ptp(arr)
 
         # GH11163
         s = Series([3, 5, np.nan, -3, 10])
-        assert s.ptp() == 13
-        assert pd.isna(s.ptp(skipna=False))
+        with tm.assert_produces_warning(FutureWarning, check_stacklevel=False):
```
[the remark this answers is not in the document]
Comment (KalyanGokhale) [reply pandas/tests/series/test_analytics.py]: Yes - it seems to be required, not for this particular check, but for e.g. the following
```python
tm.assert_series_equal(s.ptp(level=0), expected)
```
without it, this test was failing with `AssertionError: Warning not set with correct stacklevel.`

I also did trial-and-error with different stacklevels, without any success (though stacklevel=4 seems to be appropriate for where the warning is raised)

Any suggestions? Thanks

File: pandas/core/generic.py [hunk 6]
```diff
@@ -8875,13 +8875,24 @@ def _add_series_only_operations(cls):
...
+            """
+            .. deprecated:: 0.24.0
+                Use numpy.ptp instead
+
+            Returns the difference between the maximum value and the
             minimum value in the object. This is the equivalent of the
-            ``numpy.ndarray`` method ``ptp``.""",
+            ``numpy.ndarray`` method ``ptp``.
+
+            .. deprecated:: 0.24.0
```
Comment (jreback) [inline pandas/core/generic.py]: hmm this is duplicated here

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/generic.py (hunk 6 above)
Comment (KalyanGokhale) [reply pandas/core/generic.py]: This is based on comment from @jorisvandenbossche 
`This needs a blank line above this one. Can you also put the deprecated part below the text?` - waiting for his reply if the edit was as expected

Comment (KalyanGokhale) [reply pandas/core/generic.py]: My bad - misinterpreted his review and duplicated in hurry - fixing now

Comment (KalyanGokhale) [reply pandas/core/generic.py]: Done thanks

**Review by jorisvandenbossche — APPROVED** [review]

**Comment by dragoljub** [comment]
Just curious why Series.ptp() was deprecated? It was very useful shorthand for finding the period range of a DateTime column.

```python
df.DATE_TIME.ptp()

Timedelta('56 days 01:04:15')
```

**Pull request merged**